File: day7/part2/main.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ROWS 256
#define MAX_COLS 256

int main(void) {
  FILE *file = fopen("list.txt", "r");
  if (!file) {
    perror("Failed to open list.txt");
    return 1;
  }

  char grid[MAX_ROWS][MAX_COLS];
  int rows = 0;
  int cols = 0;

  while (fgets(grid[rows], MAX_COLS, file)) {

    int len = strlen(grid[rows]);
    if (len > 0 && grid[rows][len - 1] == '\n') {
      grid[rows][len - 1] = '\0';
      len--;
    }
    if (rows == 0) {
      cols = len;
    }
    rows++;
  }
  fclose(file);

  int start_col = -1;
  for (int c = 0; c < cols; c++) {
    if (grid[0][c] == 'S') {
      start_col = c;
      break;
    }
  }

  if (start_col == -1) {
    fprintf(stderr, "Could not find starting position 'S'\n");
    return 1;
  }

  long long *timelines = calloc(cols, sizeof(long long));
  long long *next_timelines = calloc(cols, sizeof(long long));

  timelines[start_col] = 1;

  for (int r = 1; r < rows; r++) {

    memset(next_timelines, 0, cols * sizeof(long long));

    for (int c = 0; c < cols; c++) {
      if (timelines[c] > 0) {

        if (grid[r][c] == '^') {

          if (c - 1 >= 0) {
            next_timelines[c - 1] += timelines[c];
          }
          if (c + 1 < cols) {
            next_timelines[c + 1] += timelines[c];
          }
        } else {

          next_timelines[c] += timelines[c];
        }
      }
    }

    long long *temp = timelines;
    timelines = next_timelines;
    next_timelines = temp;
  }

  long long total_timelines = 0;
  for (int c = 0; c < cols; c++) {
    total_timelines += timelines[c];
  }

  printf("Total timelines: %lld\n", total_timelines);

  free(timelines);
  free(next_timelines);

  return 0;
}

File: day7/part2/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func main() {
	file, err := os.Open("../list.txt")
	if err != nil {
		log.Fatalf("Failed to open list.txt: %v", err)
	}
	defer file.Close()

	var grid []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		log.Fatalf("Error reading file: %v", err)
	}

	rows := len(grid)
	if rows == 0 {
		log.Fatal("Empty grid")
	}

	cols := len(grid[0])

	// Find starting position 'S'
	startCol := -1
	for c := 0; c < cols; c++ {
		if grid[0][c] == 'S' {
			startCol = c
			break
		}
	}

	if startCol == -1 {
		log.Fatal("Could not find starting position 'S'")
	}

	// Initialize timelines
	timelines := make([]int64, cols)
	nextTimelines := make([]int64, cols)

	timelines[startCol] = 1

	// Process each row
	for r := 1; r < rows; r++ {
		// Reset next_timelines
		for i := range nextTimelines {
			nextTimelines[i] = 0
		}

		for c := 0; c < cols; c++ {
			if timelines[c] > 0 {
				if grid[r][c] == '^' {
					// Split to left and right
					if c-1 >= 0 {
						nextTimelines[c-1] += timelines[c]
					}
					if c+1 < cols {
						nextTimelines[c+1] += timelines[c]
					}
				} else {
					// Continue straight
					nextTimelines[c] += timelines[c]
				}
			}
		}

		// Swap slices
		timelines, nextTimelines = nextTimelines, timelines
	}

	// Calculate total timelines
	var totalTimelines int64
	for c := 0; c < cols; c++ {
		totalTimelines += timelines[c]
	}

	fmt.Printf("Total timelines: %d\n", totalTimelines)
}
